### PR TITLE
feat: enable iceberg table format for athena spark

### PR DESCRIPTION
**Pull request by sankeerthnagapuri**
# Description

- boto3 version update to latest
- The latest version supports SparkProperties engine configuration
- Iceberg table can only be created in Spark Athena with custom SparkProperties
- Other use cases for SparkProperties: 
   - Encryption
   - Cross account catalog querying
   - Requester pays s3 buckets

- Enhancements:
   - While waiting for a session to become IDLE, currently we raise error if we get the status of the session as "FAILED", "TERMINATED", "DEGRADED". Instead we will now be able to use a different/new session.
   - When a session is terminated/failed/degraded/timed-out, remove that session from spark_session_locks.
   - As we are introducing engine configuration into the mix, we should make sure we use the right engine configured session for the model. So, tagging the session description with dbt invocation and engine configuration hash. This way we can filter for sessions with correct configuration. Also, if the thread/session count has reached its limit and we are looking to create a new session with different configuration which is not yet available.
   - Updated logging to log the S3 URI path of STDERR file 
   - **Threads**: I have not removed the logic to get the spark_threads config but I have removed the documentation for spark_threads. The reason being the recommendation that threads should be equal to spark_threads. And that actually makes sense. When we have 8 threads and 8 spark_threads (sessions), it makes sense that all 8 can run concurrently. But if 8 threads and 4 spark_threads (sessions), the rest of 4 threads will have wait which doesnt make sense and it is exactly the same as 4 threads and 4 sessions instead. And finally in case of 8 threads and 12 spark_threads (sessions), we will have 4 additional sessions created for not really using them. So to reduce the complexity, it would be better to just use threads.
   - **Sessions**: I think we should just run the number of models concurrently based on the number of threads. And to maintain the model execution concurrency limit based of thread count, we should be able to create or reuse existing sessions. Each python model can be associated with its own engine configuration or we can have multiple models sharing same engine configuration. Based on the thread limit, we should be able to create session if an existing session configuration doesnt match with the new model. Below is the logic I followed for getting the session id:
     - Get session id
        1. Thread limit not reached: Will create a new session to make sure the thread limit concurrency is honored.
        2. Thread limit reached:
            1. Same config session in IDLE state: Will reuse this session
            2. Same config session in BUSY state: 
                1. No other query is in queue to this session (using dict: `spark_session_load`): Will wait for this session to turn IDLE and reuse this session
                2. Already a query is in queue to this session (using dict: `spark_session_load`): Will create a new session to make sure the thread limit concurrency is honored.
            3. Same config session doesn't exist: Will create a new session to make sure the thread limit concurrency is honored.
   - **Timeout**: We will use timeout to only stop calculation and not terminate session. Default 12 hours. And session timeout is based on its IDLE time period value which we hardcoded to 10 minutes, so no need for another configuration.

- Recommendations:
  - Noticed that a merge trino sql query will fail when an iceberg table is created using spark and we have not set the format_version for that iceberg table. The reason is that spark by default only creates iceberg version 1 tables and athena trino engine only works with iceberg version 2. So, recommend setting up the format version in table_properties for iceberg table configuration (See example below). Also added to README.

## Models used to test - Optional
```
def model(dbt, spark_session):
    dbt.config(
        materialized="table",
        on_schema_change="append_new_columns",
        table_type="iceberg",
        format="parquet",
        partitioned_by=['bucket(5, user_id)','status'],
        table_properties={
                    "format-version": "2"
                },
    )

    data = [("A", "pi", "active", "17.89", "1", "100000000")]
    cols = ["user_id", "name", "status", "cost", "quantity", "quantity_big"]
    df = spark_session.createDataFrame(data, cols)

    return df
```

## Checklist

- [ ] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You kept your Pull Request small and focused on a single feature or bug fix.
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: dbt/adapters/athena/session.py [hunk 1]
```diff
@@ -261,9 +298,13 @@ def release_session_lock(self, session_id: str) -> None:
...
-            LOGGER.debug(f"Terminating session: {session_id}")
-            self.athena_client.terminate_session(SessionId=session_id)
-        with self.lock:
+            LOGGER.debug(f"Terminating timed out session: {session_id}")
+            self.terminate_session(session_id)
+
+        if session_status["State"] in ["TERMINATING", "TERMINATED", "DEGRADED", "FAILED"]:
+            spark_session_locks.pop(UUID(session_id), "Session id not found")
+
+        with self.lock:  # why do we need this context manager? Why cant it be released without this?
```
Comment (sankeerthnagapuri) [inline dbt/adapters/athena/session.py:307]: @Avinash-1394  do we need this context manager? _I am not an expert with this threading library_ :-)

File: dbt/adapters/athena/constants.py [hunk 2]
```diff
@@ -6,6 +6,36 @@
 DEFAULT_SPARK_COORDINATOR_DPU_SIZE = 1
 DEFAULT_SPARK_MAX_CONCURRENT_DPUS = 2
 DEFAULT_SPARK_EXECUTOR_DPU_SIZE = 1
-DEFAULT_SPARK_SESSION_TIMEOUT = 15 * 60
+DEFAULT_CALCULATION_TIMEOUT = 43200  # seconds = 12 hours
+SESSION_IDLE_TIMEOUT_MIN = 10  # minutes
```
Comment (Avinash-1394) [inline dbt/adapters/athena/constants.py:10]: praise: Thanks for splitting the timeouts for the calculation and the session!

File: dbt/adapters/athena/constants.py [hunk 3]
```diff
@@ -6,6 +6,36 @@
...
+
+SPARKPROPERTIES = {
```
Comment (Avinash-1394) [inline dbt/adapters/athena/constants.py]: suggestion: It will be better if we rename this to `DEFAULT_SPARK_PROPERTIES` or `SPARK_PROPERTIES` just to remain consistent with the rest of the variables defined

File: dbt/adapters/athena/python_submissions.py [hunk 4]
```diff
@@ -162,18 +144,48 @@ def submit(self, compiled_code: str) -> Any:
...
+                    f"Session {self.session_id} did not become free within {self.timeout} seconds. Will try with a different session."
+                )
+                if "session_id" in self.__dict__:
+                    del self.__dict__["session_id"]
+                break
+
     def poll_until_execution_completion(self, calculation_execution_id: str) -> Any:
         """
-        Poll the status of a calculation execution until it is completed, failed, or cancelled.
+        Poll the status of a calculation execution until it is completed, failed, or canceled.
```
Comment (Avinash-1394) [inline dbt/adapters/athena/python_submissions.py:196]: praise: Good catch on the spelling for canceled! Just verified it here https://docs.aws.amazon.com/athena/latest/APIReference/API_CalculationStatus.html

File: dbt/adapters/athena/session.py [hunk 5]
```diff
@@ -102,46 +108,16 @@ def athena_client(self) -> Any:
             "athena", config=get_boto3_config(num_retries=self.credentials.effective_num_retries)
         )
 
-    def get_new_sessions(self) -> List[UUID]:
+    @cached_property
+    def session_description(self) -> str:
```
Comment (Avinash-1394) [inline dbt/adapters/athena/session.py:119]: question: Should this be a cached property? The session manager will be handling multiple sessions and as such a single session is not a property of the `AthenaSparkSessionManager`

File: dbt/adapters/athena/session.py [hunk 6]
```diff
@@ -208,13 +167,20 @@ def start_session(self) -> UUID:
             dict: The session information dictionary.
 
         """
+        description = self.session_description
```
Comment (Avinash-1394) [inline dbt/adapters/athena/session.py:185]: concern: I've questioned if the session description should be a property of the session manager. If its not supposed to be then this will have to be modified

File: dbt/adapters/athena/session.py [hunk 7]
```diff
@@ -2,18 +2,22 @@
...
-from dbt.adapters.athena.constants import DEFAULT_THREAD_COUNT, LOGGER
+from dbt.adapters.athena.constants import DEFAULT_THREAD_COUNT, LOGGER, SESSION_IDLE_TIMEOUT_MIN
+from dbt.adapters.athena.relation import AthenaRelation
 from dbt.contracts.connection import Connection
+from dbt.events.functions import get_invocation_id
 from dbt.exceptions import DbtRuntimeError
 
-spark_session_locks: Dict[UUID, threading.Lock] = {}
+invocation_id = get_invocation_id()
+spark_session_list: Dict[UUID, str] = {}
```
Comment (Avinash-1394) [inline dbt/adapters/athena/session.py:23]: concern: The reason I originally had a `threading.Lock` for each session is so that multiple dbt threads don't try to allocate the same session and get deadlocked waiting for that session when other sessions are available. How have you handled this scenario?

File: dbt/adapters/athena/python_submissions.py [hunk 8]
```diff
@@ -182,27 +194,46 @@ def poll_until_execution_completion(self, calculation_execution_id: str) -> Any:
...
+            execution_session = execution_response.get("SessionId", None)
+            execution_status = execution_response.get("Status", None)
+            execution_result = execution_response.get("Result", None)
+            execution_stderr_s3_path = ""
+            if execution_result:
+                execution_stderr_s3_path = execution_result.get("StdErrorS3Uri", None)
+
+            execution_status_state = ""
+            execution_status_reason = ""
+            if execution_status:
```
Comment (Avinash-1394) [inline dbt/adapters/athena/python_submissions.py]: suggestion (non-blocking): It would be good to log the various results here

File: dbt/adapters/athena/python_submissions.py [hunk 9]
```diff
@@ -162,18 +144,48 @@ def submit(self, compiled_code: str) -> Any:
...
+        while True:
+            timer = 0
+            session_status = self.get_current_session_status()["State"]
+            if session_status in ["TERMINATING", "TERMINATED", "DEGRADED", "FAILED"]:
+                LOGGER.debug(
+                    f"The session: {self.session_id} was not available. Got status: {session_status}. Will try with a different session."
+                )
+                self.spark_connection.remove_terminated_session(self.session_id)
+                if "session_id" in self.__dict__:
+                    del self.__dict__["session_id"]
```
Comment (Avinash-1394) [inline dbt/adapters/athena/python_submissions.py:179]: concern: I think once this property is deleted I think we might have to raise an error and fail the model. Else we somehow have to assign a new session and retry which I think might be complicated. Because otherwise I'm not sure how this model will succeed

File: dbt/include/athena/macros/materializations/models/table/table.sql [hunk 10]
```diff
@@ -98,7 +98,11 @@
         {%- endif -%}
 
         {% set query_result = safe_create_table_as(False, tmp_relation, compiled_code, language, force_batch) %}
-
+        {%- if language == 'python' -%}
```
Comment (Avinash-1394) [inline dbt/include/athena/macros/materializations/models/table/table.sql:125]: question: I think I faced some error when I tried to call the `safe_create_table_as` outside of the first call in the stack. Curious if you run into the same issue or if you handled it somehow?

File: dbt/adapters/athena/python_submissions.py
Comment (sankeerthnagapuri) [inline dbt/adapters/athena/python_submissions.py:1]: @Avinash-1394 can you please look into this workaround solution commit, i applied to address empty code submissions.

File: dbt/adapters/athena/constants.py (hunk 3 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/constants.py]: definitely will make this change

File: dbt/adapters/athena/python_submissions.py
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/python_submissions.py:1]: here is the commit: https://github.com/Avinash-1394/dbt-athena-1/pull/1/commits/fdbbab63bbd466c4d60147607205188aef3a7baf

File: dbt/include/athena/macros/materializations/models/table/table.sql (hunk 10 above)
Comment (sankeerthnagapuri) [reply dbt/include/athena/macros/materializations/models/table/table.sql:125]: the issue i encountered is when a table already exists 
for high availability reasons the process is set up to create a temp table first and rename that to main table 
but the python code was like
1. rename temp table to main table
2. create temp table

and there will be error as temp table doesnt exist in the first step 

after this change:
1. create temp table
2. rename temp table to main table

Comment (sankeerthnagapuri) [reply dbt/include/athena/macros/materializations/models/table/table.sql:125]: But other than that I didnt encounter any other issues

File: dbt/adapters/athena/python_submissions.py (hunk 9 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/python_submissions.py:179]: looks like this is a while true loop and it keeps checking for the session id
if session id doesnt exist it is attempting to get a new session id 
so deleting means to grab a new session and cache that
i actually made a couple big changes with regards to session creation and threads
let me add a comment with the detailed explaination

File: dbt/adapters/athena/python_submissions.py (hunk 8 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/python_submissions.py]: sure will check this out

File: dbt/adapters/athena/session.py (hunk 7 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:23]: let me edit the description with the explanation to the change i made for session management

Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:23]: Updated the description of this PR to include more details about threads, sessions and timeout

Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:23]: I am not an expert with threading library but with limited knowledge all that I am doing is using the self.Lock to make changes to the dictionary of session list and while accessing them to see if a session already exist or not, i am taking a copy variable of the list of session dictionary items and checking based on the description which is a hash built on engine configuration and concatenated with dbt invocation id.
Also I using this description to pass to the description field when we create a session. This helps us effectively searching for dbt invocation or engine configuration in aws console.

In short, all that i am trying to implement here is:
we will maintain a dictionary of session id and description
and we check if a new model in the run has same description (i.e., engine config) or not 
and if there is we use it if not create new one. I dont think we need listing sessions everytime rather this dictionary should help us drive the available sessions. We are also popping out terminated sessions and waiting on idle sessions. 

Please let me know if this makes sense.

File: dbt/adapters/athena/session.py (hunk 5 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:119]: it is just engine config and invocation id which distinguishes a session. So for a given session this will be its property which i am using to tag the sessions while creating them.
And i tested this out by running 11 models.
3 models with same configuration shared the same session and others split into different sessions based on their engine config.

image

Below are the sample models i used for testing this session setup:
[models.zip](https://github.com/Avinash-1394/dbt-athena-1/files/13816049/models.zip)

File: dbt/adapters/athena/constants.py (hunk 3 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/constants.py]: done 
thank you!

File: dbt/adapters/athena/python_submissions.py
Comment (Avinash-1394) [reply dbt/adapters/athena/python_submissions.py:1]: Makes sense to exit early if there is nothing to calculate. LGTM!

**Comment by Avinash-1394** [comment]
> So to reduce the complexity, it would be better to just use threads.

Can you reconsider not doing that specific change? It was raised in this comment https://github.com/dbt-athena/dbt-athena/pull/248#discussion_r1391581110 and it looks like they want additional control over how many spark sessions are created instead of relying just on threads.

From testing I've noticed that the biggest cost if often associated with spinning up new spark sessions so if people want to to trade off time for money for reusing a single spark session and making dbt threads wait that should be allowed.

File: dbt/adapters/athena/session.py (hunk 7 above)
Comment (Avinash-1394) [reply dbt/adapters/athena/session.py:23]: That does make sense. The only scenario I was afraid of was that multiple dbt threads will try to allocate the same session but it looks like using the self.Lock does alleviate that scenario. 

I'm a little unsure what `spark_session_load: Dict[UUID, int]` does. Is that to queue models into sessions that have the same config? How is the maximum load determined?

File: dbt/adapters/athena/session.py (hunk 5 above)
Comment (Avinash-1394) [reply dbt/adapters/athena/session.py:119]: Got it! That makes sense! 

Do you mind if I add those models to https://github.com/Avinash-1394/dbt-spark-development? It would be good to provide that repo as a reference to people new to athena spark.

File: dbt/adapters/athena/session.py (hunk 7 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:23]: Yeah it is just meant to manage the number of queries in queue.
I have restricted it to 1 to make sure thread concurrency is maintained (if users don't care thread concurrency, may be we can make this a config and users can set this up to whatever maximum)

But essentially it is meant to make sure no more than 1 query is in queue per session.
for ex: lets say we have 6 models and we are going with 3 threads config
model1: _configA_
model2: _configA_
model3: configB
model4: configB
model5: configB
model6: configB

With the 3 thread config, we create three sessions for the first 3 models to ensure thread concurrency of 3 is maintained (although two of the first three are sharing the same config A  we still create three sessions to maintain thread concurrency limit of 3). Now let's say all three models are completed and now to take up the next three, we only have one session with configB to run the next 3 models. Meaning we would queue 3 models to the same session and reducing the concurrency from 3 to 1. So in order to tackle this issue, I am using this spark_session_load variable to restricting to hold only 1 query per session and if there is already 1 query in queue per session, we will not hold the next two in the same session queue rather we will create 2 other sessions and maintain concurrency of 3 defined by threads config.

In short this variable ensures thread concurrency by restricting a maximum of 1 query in queue per session.

**Comment by sankeerthnagapuri** [comment]
> > So to reduce the complexity, it would be better to just use threads.
> 
> Can you reconsider not doing that specific change? It was raised in this comment [dbt-athena#248 (comment)](https://github.com/dbt-athena/dbt-athena/pull/248#discussion_r1391581110) and it looks like they want additional control over how many spark sessions are created instead of relying just on threads.
> 
> From testing I've noticed that the biggest cost if often associated with spinning up new spark sessions so if people want to to trade off time for money for reusing a single spark session and making dbt threads wait that should be allowed.

I think that makes sense to reuse the sessions but with this sparkproperties configuration coming into mix, the reuse of sessions is now restricted to reusing sessions with same config and if not then the query would fail due to misconfigured session. And we are reusing the sessions with current logic too (example image) but only as long as waiting for session doesn't drop the thread concurrency. (Please refer to this comment: https://github.com/Avinash-1394/dbt-athena-1/pull/1#discussion_r1442823370).

To increase the session reusability we can do it 2 ways:
1. We should have user set the threads to less number and if it is 1 then more reusability. But we also have to expose session idle time out config which I hardcoded to 10 minutes. If we expose and users can increase this to hours and reduce the threads to 1, meaning they will reuse the session created with a config that may be idle for some long time until another model looks for the same config session upon which that session will be used (hopefully the session idle timeout is within that time period the next model with same config gets to run). So, lesser the threads and more the idle time out, higher the chances of session reusability. But having sessions idle for so long when there would be no other model in queue also incurs additional costs I feel. Not sure of the right balance here between more idle sessions vs more sessions with less run times.
2. The other way is making `spark_session_load` constant available as config. That way users can config the number of sessions they can have in queue but increasing it more than the thread limit doesnt make any impact. The reason being, even if they configure it to be more than thread limit, the number of threads would still limit the number of models to run concurrently so at the most the queue can be as maximum as the number of models that get triggered at the same time based on thread limit. for ex: lets say we have a thread limit of 3 and we completed the first three models. Then based on the thread limit only 3 immediate models based on lineage will be eligible to run concurrently. lets say they all have same config as one of the exiting session created by one of first three models. At this point the maximum this queue can be is 3. Even if you set this to 4, we will only 3 models available for execution based on thread limit. And if one of the model is completed, we substract the count from `spark_session_load` so it will be 2 now and thread limit would also add only one new model to run thereby maintaining again at the most 3 models in queue.

File: dbt/adapters/athena/session.py (hunk 5 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:119]: Please use them but they are all same models and same config. The different one I added to readme.

File: dbt/adapters/athena/session.py (hunk 7 above)
Comment (Avinash-1394) [reply dbt/adapters/athena/session.py:23]: Very nice! 

> we will not hold the next two in the same session queue rather we will create 2 other sessions and maintain concurrency of 3 defined by threads config

Small clarification, would we terminate the sessions created with configA in this scenario? Else we would have created 5 sessions but only 3 should be in use at any time defined by the threads parameter. I'm afraid we are not controlling how many sessions are created if we go down that route. 

For example, if the threads=16 and models=64, I can envision a scenario where the first 16 use a unique model configs so the remaining models have to create a session for their execution. That is a completely fine scenario but the total sessions created could become 64 and many of them might be IDLE too long.

**Comment by Avinash-1394** [comment]
> > > So to reduce the complexity, it would be better to just use threads.
> > 
> > 
> > Can you reconsider not doing that specific change? It was raised in this comment [dbt-athena#248 (comment)](https://github.com/dbt-athena/dbt-athena/pull/248#discussion_r1391581110) and it looks like they want additional control over how many spark sessions are created instead of relying just on threads.
> > From testing I've noticed that the biggest cost if often associated with spinning up new spark sessions so if people want to to trade off time for money for reusing a single spark session and making dbt threads wait that should be allowed.
> 
> I think that makes sense to reuse the sessions but with this sparkproperties configuration coming into mix, the reuse of sessions is now restricted to reusing sessions with same config and if not then the query would fail due to misconfigured session. And we are reusing the sessions with current logic too (example image) but only as long as waiting for session doesn't drop the thread concurrency. (Please refer to this comment: [#1 (comment)](https://github.com/Avinash-1394/dbt-athena-1/pull/1#discussion_r1442823370)).
> 
> To increase the session reusability we can do it 2 ways:
> 
>     1. We should have user set the threads to less number and if it is 1 then more reusability. But we also have to expose session idle time out config which I hardcoded to 10 minutes. If we expose and users can increase this to hours and reduce the threads to 1, meaning they will reuse the session created with a config that may be idle for some long time until another model looks for the same config session upon which that session will be used (hopefully the session idle timeout is within that time period the next model with same config gets to run). So, lesser the threads and more the idle time out, higher the chances of session reusability. But having sessions idle for so long when there would be no other model in queue also incurs additional costs I feel. Not sure of the right balance here between more idle sessions vs more sessions with less run times.
> 
>     2. The other way is making `spark_session_load` constant available as config. That way users can config the number of sessions they can have in queue but increasing it more than the thread limit doesnt make any impact. The reason being, even if they configure it to be more than thread limit, the number of threads would still limit the number of models to run concurrently so at the most the queue can be as maximum as the number of models that get triggered at the same time based on thread limit. for ex: lets say we have a thread limit of 3 and we completed the first three models. Then based on the thread limit only 3 immediate models based on lineage will be eligible to run concurrently. lets say they all have same config as one of the exiting session created by one of first three models. At this point the maximum this queue can be is 3. Even if you set this to 4, we will only 3 models available for execution based on thread limit. And if one of the model is completed, we substract the count from `spark_session_load` so it will be 2 now and thread limit would also add only one new model to run thereby maintaining again at the most 3 models in queue.

Thanks for clarifying that. I think the reasoning is sound and I agree its impossible to know what the right balance is. I think it does make sense to allow configuring `spark_session_load` but we could tackle that as an improvement in another MR. This MR is quite big and I'm afraid we might introduce too many changes at once.

**Comment by Avinash-1394** [comment]
A small request, I have commented out a functional test for python submissions `tests/functional/adapter/test_python_submissions.py`. Its commented out because it doesn't pass the CI for that athena adapter repo. But can you uncomment and run that test locally to make sure it works with the current changes?

Might even be good to add additional models with the config changes from this MR into that test but that is not mandatory.

File: dbt/adapters/athena/session.py (hunk 7 above)
Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py:23]: Yeah actually that is the reason I reduced the default (20 min) idle timeout session set by AWS to a hardcoded 10 minute. 
So within 10 min there should be a new model with the same configuration available to keep it alive and if not it will be terminated.
So yes model 1 with config A will be terminated if it doesn't find a model in the next to run within 10 minutes.

Do you prefer exposing this session idle timeout as a config variable instead of us hard coding to 10 min?

**Comment by sankeerthnagapuri** [comment]
> A small request, I have commented out a functional test for python submissions `tests/functional/adapter/test_python_submissions.py`. Its commented out because it doesn't pass the CI for that athena adapter repo. But can you uncomment and run that test locally to make sure it works with the current changes?
> 
> Might even be good to add additional models with the config changes from this MR into that test but that is not mandatory.

Sure I can try that out and let you know 👍

**Comment by sankeerthnagapuri** [comment]
> > A small request, I have commented out a functional test for python submissions `tests/functional/adapter/test_python_submissions.py`. Its commented out because it doesn't pass the CI for that athena adapter repo. But can you uncomment and run that test locally to make sure it works with the current changes?
> > Might even be good to add additional models with the config changes from this MR into that test but that is not mandatory.
> 
> Sure I can try that out and let you know 👍

Tested by uncommenting the file: `tests/functional/adapter/test_python_submissions.py`
Below is the summary:

```
16 workers [98 items]     
sssssssssssss.........sss.ss....................x.................................................    [100%]
============================================= warnings summary =============================================
tests/functional/adapter/test_docs.py: 9 warnings
tests/functional/adapter/test_basic_iceberg.py: 18 warnings
tests/functional/adapter/test_basic_hive.py: 31 warnings
tests/functional/adapter/test_basic_hive_native_drop.py: 15 warnings
tests/functional/adapter/test_basic_iceberg_native_drop.py: 6 warnings
tests/functional/adapter/test_incremental_iceberg.py: 39 warnings
tests/functional/adapter/test_change_relation_types.py: 20 warnings
tests/functional/adapter/test_constraints.py: 3 warnings
tests/functional/adapter/test_detailed_table_type.py: 3 warnings
tests/functional/adapter/test_force_batch.py: 9 warnings
tests/functional/adapter/test_incremental_iceberg_merge_no_updates.py: 3 warnings
tests/functional/adapter/utils/test_utils.py: 75 warnings
tests/functional/adapter/test_python_submissions.py: 6 warnings
tests/functional/adapter/test_partitions.py: 15 warnings
tests/functional/adapter/test_seed_by_insert.py: 8 warnings
tests/functional/adapter/test_quote_seed_column.py: 7 warnings
tests/functional/adapter/test_snapshot.py: 36 warnings

=================== 79 passed, 18 skipped, 1 xfailed, 303 warnings in 917.95s (0:15:17) ==================== 
```

**Comment by Avinash-1394** [comment]
@sankeerthnagapuri It looks like some pre commit checks are failing https://github.com/Avinash-1394/dbt-athena-1/actions/runs/7469823790/job/20365562831?pr=1. Would you be able to install pre-commit hooks and fix those?

**Comment by sankeerthnagapuri** [comment]
> @sankeerthnagapuri It looks like some pre commit checks are failing https://github.com/Avinash-1394/dbt-athena-1/actions/runs/7469823790/job/20365562831?pr=1. Would you be able to install pre-commit hooks and fix those? 

@Avinash-1394 done fixed the precommit hooks

File: dbt/adapters/athena/session.py [hunk 11]
```diff
@@ -41,7 +49,14 @@ class AthenaSparkSessionManager:
...
     """
 
-    def __init__(self, credentials: Any, timeout: int, polling_interval: float, engine_config: Dict[str, int]) -> None:
+    def __init__(
+        self,
+        credentials: Any,
+        timeout: int,
+        polling_interval: float,
+        engine_config: Dict[str, int],
+        relation_name: str | None = None,
```
Comment (Avinash-1394) [inline dbt/adapters/athena/session.py]: This `| None ` is causing an issue with the unit test suite some how

Comment (sankeerthnagapuri) [reply dbt/adapters/athena/session.py]: @Avinash-1394 
Just got to know that: str | None syntax is only supported in 3.10 or later
Defaulted to string "N/A" as this input variable I added is only for logging purpose

**Comment by sankeerthnagapuri** [comment]
@Avinash-1394 fixed the test_config.py assertion
Also removed tests associated with functions which no longer exist

**Comment by Avinash-1394** [comment]
@sankeerthnagapuri There seems to be one small error with the pre-commit. Sorry for all the back and forth and I hope this is not too annoying!

**Comment by sankeerthnagapuri** [comment]
My bad, sorry let me fix it right away

**Comment by sankeerthnagapuri** [comment]
> @sankeerthnagapuri There seems to be one small error with the pre-commit. Sorry for all the back and forth and I hope this is not too annoying!

@Avinash-1394 fixed the precommit.

**Comment by sankeerthnagapuri** [comment]
@Avinash-1394 can you please help me understand the error with unit test: test_python_submissions
in my local i see it takes forever at this step: ```tests\unit\test_python_submissions.py .xx```

**Comment by Avinash-1394** [comment]
Do you mean `tests/functional/adapter/test_python_submissions.py`? That one actually tries to run the models specified in that file. You can actually check if the sessions are getting created and the models are running from the Athena UI. I will try cloning the branch and running that test and provide an output.

**Comment by sankeerthnagapuri** [comment]
Nah, I meant unit test not functional test. Looks like that one failed I guess right? 
I am not familiar with writing tests. So, I might take some time to understand the exact issue we have with the current unit test for python submissions. 
Will spend some time tomorrow and figure it out unless u know the issue.

**Comment by sankeerthnagapuri** [comment]
@Avinash-1394 tested units tests and pre-commit 
passed all checks
Please run the CI workflow.

**Comment by Avinash-1394** [comment]
@sankeerthnagapuri All green! Merged 👍🏽
